cassandra-diagnostics-connector21/src/main/java/io/smartcat/cassandra/diagnostics/connector/DiagnosticsQueryHandler.java

```diff
@@ -0,0 +1,125 @@
+package io.smartcat.cassandra.diagnostics.connector;
+
+import org.apache.cassandra.cql3.BatchQueryOptions;
+import org.apache.cassandra.cql3.CQLStatement;
+import org.apache.cassandra.cql3.QueryHandler;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.QueryProcessor;
+import org.apache.cassandra.cql3.statements.BatchStatement;
+import org.apache.cassandra.cql3.statements.ParsedStatement;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.exceptions.RequestExecutionException;
+import org.apache.cassandra.exceptions.RequestValidationException;
+import org.apache.cassandra.service.QueryState;
+import org.apache.cassandra.transport.messages.ResultMessage;
+import org.apache.cassandra.transport.messages.ResultMessage.Prepared;
+import org.apache.cassandra.utils.MD5Digest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.smartcat.cassandra.diagnostics.Diagnostics;
+
+/**
+ * Custom diagnostics query handler which hooks into query path and reports queries.
+ */
+public class DiagnosticsQueryHandler implements QueryHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DiagnosticsQueryHandler.class);
+
+    private static final String INITIALIZATION_THREAD_NAME = "cassandra-diagnostics-agent";
+
+    private final QueryProcessor queryProcessor = QueryProcessor.instance;
+
+    private static Connector21QueryReporter queryReporter;
+
+    /**
+     * Create diagnostics query handler and init diagnostics.
+     */
+    public DiagnosticsQueryHandler() {
+        initDiagnostics();
+    }
+
+    /**
+     * Create and activate diagnostics as part of this query handler.
+     */
+    private void initDiagnostics() {
+        LOGGER.info("Cassandra Diagnostics starting.");
+        final Diagnostics diagnostics = new Diagnostics();
+        Thread th = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                diagnostics.activate();
+                LOGGER.info("Cassandra Diagnostics initialized.");
+            }
+        });
+        th.setName(INITIALIZATION_THREAD_NAME);
+        th.setDaemon(true);
+        th.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+            @Override
+            public void uncaughtException(Thread t, Throwable e) {
+                LOGGER.error(e.getMessage(), e);
+            }
+        });
+        th.start();
+        queryReporter = new Connector21QueryReporter(diagnostics, diagnostics.getConfiguration().connector);
+    }
+
+    @Override
+    public ResultMessage process(String query, QueryState state, QueryOptions options)
+            throws RequestExecutionException, RequestValidationException {
+        LOGGER.trace("Intercepted process");
+
+        // Code from {@link org.apache.cassandra.cql3.QueryProcessor}
```

Comment:
We need parameters from two methods, one with string and one with prepared statement. Only String one is part of QueryHandler interface which we can use.

We can probably do that if we do prepare statement twice, one in queryProcessor (delegated method) and one for keyspace and table data, but I did not want to do it twice.